DOCS-1160: Document generic component (#1897)
viamrobotics/docs: docs/components/generic/_index.md

---
title: "Generic Component"
linkTitle: "Generic"
childTitleEndOverwrite: "Generic Component"
weight: 55
type: "docs"

> description: "Use a generic component to implement a component that does not fit any of the other APIs."

Reviewer: This is more in line with the descriptions of the other components - see components top level page:
```suggestion
description: "A component that does not fit any of the other APIs."
```

tags: ["generic", "components"]
icon: "/icons/components.png"

> images: ["/icons/components.png"]

Reviewer: Please ask Cristina if she can make a generic component icon

no_list: true
# SMEs:
---

The _generic_ component type is for custom components that do not fit into any of the other component {{< glossary_tooltip term_id="subtype" text="subtypes" >}}.

There are no built-in generic component models (other than `fake`).
Generic is only for [modular resource](/extend/modular-resources/) models that don't fit into any of the typical categories.

Use cases:

> - You want to implement a new model of component as a modular resource, and your device isn't an arm, a base, a board, a camera, or any of the other [built-in Viam component types](/components/).

Reviewer: You say this three times (line 15, 18, and 22) - please change this intro paragraph and condense that to once

Author: Okay I reworked a bit--still says it 1.5 times but one is about API and one is about hardware. I kinda want to say it a couple of different ways since subtypes can be so confusing for people new to Viam. LMK if you have suggestions.

- You are using someone else's modular resource, which they implemented with the generic subtype.

{{% alert title="Important" color="note" %}}

The generic component API only supports the `DoCommand` method.
If you use the generic subtype, your module needs to define any and all component functionality and pass it through `DoCommand`.

Whenever possible, it is best to use an [existing component API](/components/) instead of generic so that you do not have to replicate code.
If you want to use most of an existing API but need just a few other functions, try using the `DoCommand` endpoint and extra parameters to add custom functionality to an existing subtype.

{{% /alert %}}

## Configuration

> The generic component supports the following models.

Reviewer: This may have to change in the template
```suggestion
For configuration information, click on one of the supported generic model:
```

Author: Yeah, also not all of them are standard still.

> The generic component supports the following models.
> Click on the component model name for detailed configuration information.

Author: ```suggestion
For configuration information, click on one of the supported generic models:
```

If you want to use another generic model with Viam, you can [define a custom component](../../extend/).

## Control your board with Viam's client SDK libraries

To get started using Viam's SDKs to connect to and control your robot, go to your robot's page on [the Viam app](https://app.viam.com), navigate to the **Code sample** tab, select your preferred programming language, and copy the sample code generated.

{{% snippet "show-secret.md" %}}

When executed, this sample code will create a connection to your robot as a client.
Then control your robot programmatically by getting your `generic` component from the robot with `FromRobot` and adding API method calls, as shown in the following examples.

These examples assume you have a board called "my_board" configured as a component of your robot.
If your board has a different name, change the `name` in the code.

Be sure to import the generic package for the SDK you are using:

{{< tabs >}}
{{% tab name="Python" %}}

```python
from viam.components.generic import Generic
```

{{% /tab %}}
{{% tab name="Go" %}}

```go
import (
  "go.viam.com/rdk/components/generic"
)
```

{{% /tab %}}
{{< /tabs >}}

## API

> The generic component supports the following methods:

Reviewer: ```suggestion
The generic component supports the following method:
```

Author: I was wondering whether I should keep it uniform, and in case we ever add more methods (or document more of the general methods) and forget to change this. Will apply.
